src/tools/immutable_object_cache/ObjectCacheStore.cc

```diff
@@ -178,25 +231,23 @@ int ObjectCacheStore::handle_promote_callback(int ret, bufferlist* read_buf,
   return ret;
 }
 
-int ObjectCacheStore::lookup_object(std::string pool_nspace,
-                                    uint64_t pool_id, uint64_t snap_id,
+int ObjectCacheStore::lookup_object(std::string pool_nspace, uint64_t pool_id,
+                                    uint64_t snap_id, uint64_t object_size,
                                     std::string object_name,
                                     bool return_dne_path,
                                     std::string& target_cache_file_path) {
   ldout(m_cct, 20) << "object name = " << object_name
                    << " in pool ID : " << pool_id << dendl;
 
-  int pret = -1;
   std::string cache_file_name = get_cache_file_name(pool_nspace, pool_id, snap_id, object_name);
 
   cache_status_t ret = m_policy->lookup_object(cache_file_name);
 
   switch (ret) {
     case OBJ_CACHE_NONE: {
-      pret = do_promote(pool_nspace, pool_id, snap_id, object_name);
-      if (pret < 0) {
-        lderr(m_cct) << "fail to start promote" << dendl;
-      }
+      Cache_Object *cache_object = new Cache_Object(pool_id, snap_id,
+        object_size, object_name, pool_nspace);
+      m_promote_queue.push(cache_object);
```

Comment:
I still think this is backwards. Right now the cache policy is saying "promote this object" but if you are stuck behind QoS, you then will start to get backlogged (`m_promote_queue` will grow w/o bounds). I think the QoS logic needs to be in the policy so it can decide that it wants to promote an object because (1) it has enough space and (2) it has enough QoS available.

Reply:
OK, so I will delete the thread code. Combined with your previous idea, I think a new state "OBJ_CACHE_THROTTLE" can be added to indicate that the cache has not been cached and can not be promoted at present because of throttle. Then implement the throttle mechanism in simplepolicy.cc. Do you think this is OK?